### ISSUE TITLE
simEnv_par() function

### ISSUE THREAD

**Issue by XuejianXiong** [euba/BacArena#161]
Hi there,

I'm trying to replace simEnv() with simEnv_par() to parallel my simulation. In other words, I only changed this one command in my script. 
My script was running good with simEnv(). But now I need to run a simulation for 96 simulation hours. So I would like to use simEnv_par() to speed the simulation. However I got the following error message:

"Error in eval(substitute(expr), data, enclos = parent.frame()) : 
  invalid 'envir' argument of type 'character'
Calls: <Anonymous> ... unlist -> sapply -> lapply -> FUN -> with.default -> eval
In addition: Warning message:
In parallel::mclapply(groups, function(g) { :
  all scheduled cores encountered errors in user code"

Could you please help me to solve this problem? Thanks!

Best,
Xuejian

**Comment by Shriram369** [comment]
While using parallel version of simEnv i.e. simEnv_par(), I got the same error with the example code:

**data(Ec_core, envir = environment()) 
bac <- Bac(Ec_core,deathrate=0.05, minweight=0.05,growtype="exponential") 
arena <- Arena(n=20,m=20) 
arena <- addOrg(arena,bac,amount=10) 
arena <- addSubs(arena,40)
eval <- simEnv_par(arena,5, cluster_size = 5)** 

Error message is as follow:
Error in eval(substitute(expr), data, enclos = parent.frame()) : 
  invalid 'envir' argument of type 'character'

I could be doing it wrong. And it would be good to have your advise/ suggestion on it.

Thanks,
Shriram

**Comment by jotech** [comment]
Hi Shiram, unfortunately the parallel version of bacarena is not working right now, so I would not recommend to use it. ``simEnv_par()`` produces a lot of overhead so that the gain of parallelisation isn't enough. 
I don't have time for a better implementation at the moment, but maybe you can try to have multiple runs of bacarena for each sample or replicate? At least for me this works quite fine in parallel on a comuter cluster!

**Comment by Shriram369** [comment]
Thank you for your reply. I can confirm that multiple runs of bacarena runs fine for me, too. Cheers.

**Comment by jotech** [comment]
hi @XuejianXiong, may I ask what linear solver do you use? In the past, I encountered some instabilities in parallel jobs while using glpk but cplex was worked fine.

**Comment by jotech** [comment]
Thanks for your reply! This looks reasonable indeed. Do you have the opportunity to try and install the R package cplexAPI? For this cplex solver from IBM is need but it is free for academic users. In the past, I encountered instabilities in bigger and longer simulations on a computer cluster that were related to glpkAPI. Maybe you can give it a try?

**Comment by XuejianXiong** [comment]
I see. I did install glpkAPI for bacarena. Is the cplexAPI from here: https://cran.r-project.org/web/packages/cplexAPI/index.html? Thanks!

**Comment by jotech** [comment]
yes this is the R package but you have have IBM ILOG CPLEX installed in order to use the cplex solver in R

**Comment by XuejianXiong** [comment]
Hi @jotech,
My cluster administrator asked me to point him to the download location of "IBM ILOG CPLEX". Could you please help me with this? 
Thanks!

**Comment by jotech** [comment]
https://community.ibm.com/community/user/datascience/blogs/xavier-nodet1/2020/07/09/cplex-free-for-students
hope it is working for you!

**Comment by XuejianXiong** [comment]
Thank you @jotech!
I registered IBM yesterday and am waiting for the download link.

**Comment by XuejianXiong** [comment]
Hi @jotech, 
The CPLEX works very well! Now it only takes 4 hours to finish my job. Thank you very much!

Another question is: 
Can I add an extra enzyme, amylase 3.2.1.1, into the BacArean system using addSubs() ?
Thank you!

Best,
Xuejian

**Comment by jotech** [comment]
i'm happt that it worked for you :) 
I created a [new issue](https://github.com/euba/BacArena/issues/165) for the question concerning the external enzyme, hope this is fine for you.
otherwise please reopen this issue again
